compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/BallerinaSemanticModel.java

```diff
@@ -131,6 +133,22 @@ public BallerinaSemanticModel(BLangPackage bLangPackage, CompilerContext context
         return compiledSymbols;
     }
 
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Optional<BallerinaTypeDescriptor> getType(String fileName, LineRange range) {
+        BLangCompilationUnit compilationUnit = getCompilationUnit(fileName);
+        NodeFinder nodeFinder = new NodeFinder();
+        BLangNode node = nodeFinder.lookup(compilationUnit, range);
+
+        if (node == null) {
+            return Optional.empty();
+        }
+
+        return Optional.ofNullable(TypesFactory.getTypeDescriptor(node.type));
```

Comment:
Why we are returning NULLs in`TypesFactory.getTypeDescriptor(node.type)` ? shouldn't it return an optional type-desc ?

Reply:
We have kept the factories to return the relevant instance or `null`. Returning an `Optional` value straight from the factory may not be desirable for some of the uses of it. e.g., populating a list of types (in the union for instance). If we return `Optional` from the factory, then the list would be a `null` allowing list: `List<Optional<BallerinaTypeDescriptor>>`

Reply:
When populating a list of types still we call `TypesFactory.getTypeDescriptor()` per each type right? I am not sure how usage of `null` avoid a `null` allowing list. Idea of using `Optional<T>` for a function-return is to explicitly notify that the value can be present or not. Otherwise, API user deliberately need to go and check the function implementation. Since this is a public API I still believe `Optional<BallerinaTypeDescriptor>` is the way to go. WDYT @nadeeshaan @SupunS  ?

Reply:
I think, what @rasika says is valid if we are exposing this particular API to the public. IIRC we had to expose it to public due to the project API limitation. I would say, as long as we hide it as a public API, this is ok. WDYT rasika?